oak/server/rust/oak_runtime/src/node/http/mod.rs

```diff
@@ -0,0 +1,520 @@
+                error!(
+                    "Couldn't write message to the Http request channel: {:?}",
+                    err
+                );
+            })
+    }
+
+    fn send_invocation(
+        &self,
+        runtime: &RuntimeProxy,
+        invocation_channel: oak_abi::Handle,
+    ) -> Result<(), ()> {
+        // Create an invocation message and attach the method-invocation specific channels to it.
+        //
+        // This message should be in sync with the [`oak::http::Invocation`] from the Oak SDK:
+        // the order of the `request_reader` and `response_writer` must be consistent.
+        let invocation = crate::NodeMessage {
+            data: vec![],
+            handles: vec![self.request_reader, self.response_writer],
+        };
+
+        // Send an invocation message (with attached handles) to the Oak Node.
+        runtime
+            .channel_write(invocation_channel, invocation)
+            .map_err(|error| {
+                error!("Couldn't write gRPC invocation message: {:?}", error);
+            })
+    }
+
+    // Close all local handles except for the one that allows reading responses.
+    fn close(&self, runtime: &RuntimeProxy) {
+        if let Err(err) = runtime.channel_close(self.request_writer) {
+            error!(
+                "Failed to close request writer channel for invocation: {:?}",
+                err
+            );
+        }
+        if let Err(err) = runtime.channel_close(self.request_reader) {
+            error!(
+                "Failed to close request reader channel for invocation: {:?}",
+                err
+            );
+        }
+        if let Err(err) = runtime.channel_close(self.response_writer) {
+            error!(
+                "Failed to close response writer channel for invocation: {:?}",
+                err
+            );
+        }
+    }
+}
+
+fn get_oak_label(req: &HttpRequest) -> Result<Label, OakStatus> {
+    match req.headers.get(oak_abi::OAK_LABEL_HTTP_KEY) {
+        Some(label) => {
+            let label_bytes = bytes::Bytes::copy_from_slice(&label[..]);
+            Label::decode(label_bytes).map_err(|err| {
+                warn!("Could not parse Http label: {}", err);
+                OakStatus::ErrInternal
+            })
+        }
+        None => {
+            warn!("No HTTP label found:");
+            Err(OakStatus::ErrInternal)
+        }
+    }
+}
+
+struct HttpResponseIterator {
+    runtime: RuntimeProxy,
+    response_reader: oak_abi::Handle,
+}
+
+impl HttpResponseIterator {
+    fn read_response(&self) -> Option<HttpResponse> {
+        let read_status = self
+            .runtime
+            .wait_on_channels(&[self.response_reader])
+            .map_err(|error| {
+                error!("Couldn't wait on the HTTP response channel: {:?}", error);
+            })
+            .ok()?;
+
+        if read_status[0] == ChannelReadStatus::ReadReady {
+            match self.runtime.channel_read(self.response_reader) {
```

Comment:
This feels like a reimplementation of Receiver:

https://github.com/project-oak/oak/blob/13c544122aaa25f4818c6a47f28818108498c245/oak/server/rust/oak_runtime/src/io.rs#L82-L100